packages/main/src/plugin/kubernetes-exec-transmitter.ts

```diff
@@ -0,0 +1,116 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import type { WritableOptions } from 'node:stream';
+import { Readable, Writable } from 'node:stream';
+
+export const DEFAULT_COLUMNS: number = 80;
+export const DEFAULT_ROWS: number = 60;
+
+export interface TerminalSize {
+  height: number;
+  width: number;
+}
+
+export class ExecStreamWriter extends Writable {
+  protected transmitter: Writable;
+
+  constructor(transmitter: Writable, options?: WritableOptions) {
+    super(options);
+    this.transmitter = transmitter;
+  }
+
+  _write(chunk: unknown, encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
+    this.transmitter._write(chunk, encoding, callback);
+  }
+}
+
+export class ResizableTerminalWriter extends ExecStreamWriter {
+  protected columns: number;
+  protected rows: number;
+
+  constructor(
+    transmitter: Writable,
+    terminalSize: TerminalSize = { width: DEFAULT_COLUMNS, height: DEFAULT_ROWS },
+    options?: WritableOptions,
+  ) {
+    super(transmitter, options);
+    this.columns = terminalSize.width;
+    this.rows = terminalSize.height;
+  }
+
+  _write(chunk: unknown, encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
+    super._write(chunk, encoding, callback);
+  }
+
+  resize(terminalSize: TerminalSize): void {
+    this.columns = terminalSize.width;
+    this.rows = terminalSize.height;
+
+    this.emit('resize');
+  }
+
+  getDimension(): TerminalSize {
+    return {
+      width: this.columns,
+      height: this.rows,
+    } as TerminalSize;
+  }
+}
+
+export class BufferedStreamWriter extends Writable {
+  private readonly transmitFn: (data: Buffer) => void;
+
+  constructor(transmitFn: (data: Buffer) => void, options?: WritableOptions) {
+    super(options);
+    this.transmitFn = transmitFn;
+  }
+
+  _write(chunk: Buffer, encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
+    this.transmitFn(chunk);
+    callback();
+  }
+}
+
+export class StringLineReader extends Readable {
+  private dataQueue: string[] = [];
+  private isReading: boolean = false;
+
+  constructor() {
+    super();
+  }
+
+  readLine(data: string): void {
+    this.dataQueue.push(data);
+    if (!this.isReading) {
+      this.read();
+    }
+  }
+
+  _read(): void {
+    this.isReading = true;
+    const data = this.dataQueue.shift();
+
+    if (data) {
+      this.push(data);
+    } else {
+      this.isReading = false;
+      this.push(undefined);
+    }
+  }
+}
```

Comment:
I'm not sure to understand te Transmitter

for terminal resize what we did was
https://github.com/containers/podman-desktop/pull/3955/files

Reply:
Yes, this approach works, when we have separate method for stdout resize, like dockerode has, but client-node library doesn't have separate method to change the stdout dimensions. When we create the exec connection, we have to pass streams for stdout, stderr and stdin:

https://github.com/kubernetes-client/javascript/blob/3610c1b20bccb08aeca3aa3d37ad0fd5c1fbc425/src/exec.ts#L37-L39

this is the first point why there were created [fictitious streams](https://github.com/containers/podman-desktop/pull/5975/files#diff-8799a75d3c2928c9fe5136fceab66ed0edad4de3a675f9f5f6d3bc3c61f3dd7bR1409-R1411), which consume the handlers for [input](https://github.com/containers/podman-desktop/pull/5975/files#diff-8799a75d3c2928c9fe5136fceab66ed0edad4de3a675f9f5f6d3bc3c61f3dd7bR1397) and [output](https://github.com/containers/podman-desktop/pull/5975/files#diff-8799a75d3c2928c9fe5136fceab66ed0edad4de3a675f9f5f6d3bc3c61f3dd7bR1394-R1395) operations. Then, mandatory condition is to implement this [interface](https://github.com/kubernetes-client/javascript/blob/master/src/terminal-size-queue.ts#L3-L7) in stdout stream which can handle resize event and has dimention values. And after that, by calling [resize](https://github.com/containers/podman-desktop/pull/5975/files#diff-8799a75d3c2928c9fe5136fceab66ed0edad4de3a675f9f5f6d3bc3c61f3dd7bR1446) on stream, we can fire [event](https://github.com/containers/podman-desktop/pull/5975/files#diff-e4eec56465f6b765285da8c5467d1e766c0bf93df11693904e792927b31c5d50R65) which will be [handled](https://github.com/kubernetes-client/javascript/blob/master/src/terminal-size-queue.ts#L28) in client-node library and pass the resize event through websocket to the container. That's why we should have a custom stream. To easier support the streams in future, I wrapped them as decorators.